firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,148 @@
+package com.google.firebase.perf.application;
+
+import android.app.Activity;
+import android.util.SparseIntArray;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.core.app.FrameMetricsAggregator;
+
+import com.google.android.gms.common.util.VisibleForTesting;
+import com.google.firebase.perf.logging.AndroidLogger;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator.PerfFrameMetrics;
+import com.google.firebase.perf.util.Constants;
+
+import java.util.WeakHashMap;
+
+/**
+ * Provides FrameMetrics data from an Activity's Window. Encapsulates FrameMetricsAggregator.
+ * <p>
+ * Note: each recorder holds a reference to an Activity, so it is very important to dereference
+ * each recorder before the associated Activity is destroyed
+ * </p>
+ * */
+public class FrameMetricsRecorder {
+  private static final AndroidLogger logger = AndroidLogger.getInstance();
+  private final Activity activity;
+  private final FrameMetricsAggregator fma;
+  private final WeakHashMap<Object, PerfFrameMetrics> subTraceMap;
+  private boolean isRecording;
+
+  /**
+   * Creates a recorder for a specific activity
+   * @param activity the activity that the recorder is collecting data from
+   */
+  public FrameMetricsRecorder(Activity activity) {
+    this(activity, new FrameMetricsAggregator(), new WeakHashMap<>(), false);
+  }
+
+  @VisibleForTesting
+  FrameMetricsRecorder(Activity activity, FrameMetricsAggregator frameMetricsAggregator, WeakHashMap<Object, PerfFrameMetrics> subTraceMap, boolean isRecording) {
+    this.activity = activity;
+    this.fma = frameMetricsAggregator;
+    this.subTraceMap = subTraceMap;
+    this.isRecording = isRecording;
+  }
+
+  /**
+   * Starts recording FrameMetrics for the activity window
+   */
+  public void start() {
+    if (isRecording) {
+      logger.error("FrameMetricsAggregator is already recording %s", activity.getClass().getSimpleName());
+      return;
+    }
+    fma.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window
+   * @return FrameMetrics accumulated during the current recording
+   */
+  public PerfFrameMetrics stop() {
+    if (!isRecording) {
+      logger.error("Cannot stop because no recording was started");
+      return null;
+    }
+    isRecording = false;
+    PerfFrameMetrics data = this.snapshot();
+    try {
+      fma.remove(activity);
+    } catch (IllegalArgumentException ignored) {
+      logger.error("View not hardware accelerated. Unable to collect FrameMetrics.");
+    }
+    fma.reset();
+    return data;
+  }
+
+  /**
+   * Starts a sub-trace in the current recording
+   * @param key a UI state to associate this sub-trace with. e.g.) fragment
+   */
+  public void startSubTrace(Object key) {
```

Comment:
Since this is a generic `Object` type, can we type check internally on what the type of the object is before we start recording the metrics for the Object?

Reply:
I make it generic `Object` to be more robust and not just for fragments. Perhaps in the future some sort of Jetpack Compose UI can be associated with a sub-trace, this implementation isn't only allowing fragments, it's to be very generic and more in the perspective of `FrameMetricsAggregator`. FMA doesn't care what you are tagging this subtrace as. From FMA's perspective, we are making a sub-trace (snapshot B - snapshot A), and just giving it a "key" as a name.

Reply:
IMO that would couple the Recorder with other concepts unnecessarily. Recorder is a wrapper of FMA, and it should only be responsible for operations related to FMA. A "sub-trace" only needs a key, it doesn't matter what type it is. It could be any UI object representing that "the activity currently has X UI on screen".

Reply:
My 2c is to try to avoid using `Object` and other generic types as much as possible. It makes it easy to misuse APIs. go/java-practices/methods#types